Add buildhost VM

calancha opened the pull request:
## What does this PR change?
Add a new VM 'buildhost'

The idea is to have one VM for the 'salt minion' and another VM
for the 'Docker and Kiwi build host' in the Cucumber test suite.

This change adds flexibility to the test suite; by decoupling the 'salt minion'
from the build host we a are able to test more minion versions (not just
those with a build host functionality available).

Another advantage is a reduction in the number of clean up scenarios
(the slow readd entitlements).
See https://github.com/SUSE/spacewalk/issues/10680

- modules/controller/variables.tf
  Add buildhost configuration.

- modules/cucumber_testsuite/main.tf
  Add min-build module section.
  Remove avahi_reflector from the normal minion; keep it only for the min-build.

- salt/controller/bashrc:
  Export environment variable with the build host hostname.

- modules/controller/main.tf
  Read build host configuration.

- README_TESTING.md (Advanced deployments):
  Include the new VM in the example.

Bischoff commented on README_TESTING.md:
````diff
@@ -28,7 +28,7 @@ module "cucumber_testsuite" {
 ```
 
-This will create a test server, client and minion instances, plus a coordination node called a `controller` which runs the testsuite.
+This will create a test server, client, minion and buildhost instances, plus a coordination node called a `controller` which runs the testsuite.
````
", and"

The build host is a client, after all - maybe we don't need to specify here.

Bischoff commented on README_TESTING.md:
```diff
@@ -105,6 +105,9 @@ host_settings = {
   min-sles12sp4 = {
     present = true
   }
+  bldh-sles12sp4 = {
```
`min-build` instead of `bldh-sles12sp4`, like `min-kvm` below.

As a side note, `sles12sp4` is an error, our minions have no reason to be SLES 12 SP4. @rjmateus is going to rectify these names introduced during the refactoring.

Bischoff commented on README_TESTING.md:
````diff
@@ -123,7 +126,7 @@ host_settings = {
 ```
 
-The default value for host_settings block has a SLES12SP4 Traditional Client and SLES12SP4 Minion present:
+The default value for host_settings block has a SLES12SP4 Traditional Client, a SLES12SP4 Minion and a SLES12SP4 Docker/Kiwi present:
````
No, I think the build host should be optional.

BTW, according to `modules/controller/main.tf`below, you made it optional.

Bischoff commented on README_TESTING.md:
```diff
@@ -133,6 +136,9 @@ host_settings = {
   min-sles12sp4 = {
     present = true
   }
+  bldh-sles12sp4 = {
```
see above about its name

Bischoff commented on modules/controller/main.tf:
```diff
@@ -36,6 +36,7 @@ module "controller" {
     proxy         = var.proxy_configuration["hostname"]
     client        = length(var.client_configuration["hostnames"]) > 0 ? var.client_configuration["hostnames"][0] : null
     minion        = length(var.minion_configuration["hostnames"]) > 0 ? var.minion_configuration["hostnames"][0] : null
+    buildhost     = length(var.buildhost_configuration["hostnames"]) > 0 ? var.buildhost_configuration["hostnames"][0] : null
```
by analogy with `kvm_host` below, I would rather use `build_host` with an underscore.

Bischoff commented on modules/controller/variables.tf:
```diff
@@ -54,6 +54,14 @@ variable "minion_configuration" {
   }
 }
 
+variable "buildhost_configuration" {
```
yes

Bischoff commented on modules/controller/variables.tf:
```diff
@@ -54,6 +54,14 @@ variable "minion_configuration" {
+variable "buildhost_configuration" {
+  description = "use module.<BUILDHOST_NAME>.configuration, see main.tf.libvirt-testsuite.example"
+  default = {
+    ids       = []
```
do we need the ids ? I don't think so

Bischoff replied:
not addressed yet

Bischoff commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -125,6 +125,26 @@ module "min-sles12sp4" {
   provider_settings = lookup(local.provider_settings_by_host, "min-sles12sp4", {})
 }
 
+module "bldh-sles12sp4" {
```
see above about name

Bischoff commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -125,6 +125,26 @@ module "min-sles12sp4" {
+module "bldh-sles12sp4" {
+  source = "../minion"
+
+  quantity           = contains(local.hosts, "bldh-sles12sp4") ? 1 : 0
```
same

Bischoff commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -125,6 +125,26 @@ module "min-sles12sp4" {
+  quantity           = contains(local.hosts, "bldh-sles12sp4") ? 1 : 0
+  base_configuration = module.base.configuration
+  product_version    = var.product_version
+  image              = lookup(local.images, "bldh-sles12sp4", "sles12sp4")
```
same

Bischoff commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -125,6 +125,26 @@ module "min-sles12sp4" {
+  base_configuration = module.base.configuration
+  product_version    = var.product_version
+  image              = lookup(local.images, "bldh-sles12sp4", "sles12sp4")
+  name               = lookup(local.names, "bldh-sles12sp4", "bldh-sles12")
```
same

Bischoff commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -125,6 +125,26 @@ module "min-sles12sp4" {
+  auto_connect_to_master  = false
+  use_os_released_updates = true
+  ssh_key_path            = "./salt/controller/id_rsa.pub"
+  avahi_reflector         = var.avahi_reflector
```
right - please remove the `avahi_reflector` declaration from the normal minion declaration. It will be needed only for the build host.

Bischoff replied:
not addressed yet

Bischoff commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -219,6 +239,7 @@ module "ctl" {
   proxy_configuration     = contains(local.hosts, "pxy") ? module.pxy.configuration : { hostname = null }
   client_configuration    = contains(local.hosts, "cli-sles12sp4") ? module.cli-sles12sp4.configuration : { hostnames = [], ids = [] }
   minion_configuration    = contains(local.hosts, "min-sles12sp4") ? module.min-sles12sp4.configuration : { hostnames = [], ids = [] }
+  buildhost_configuration = contains(local.hosts, "bldh-sles12sp4") ? module.bldh-sles12sp4.configuration : { hostnames = [], ids = [] }
```
see remarks about names.

Bischoff commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -245,6 +266,7 @@ output "configuration" {
     pxy = module.pxy.configuration
     cli-sles12sp4 = module.cli-sles12sp4.configuration
     min-sles12sp4 = module.min-sles12sp4.configuration
+    bldh-sles12sp4 = module.bldh-sles12sp4.configuration
```
same

Bischoff commented on modules/cucumber_testsuite/variables.tf:
```diff
@@ -59,6 +59,9 @@ variable "host_settings" {
     min-sles12sp4 = {
       present = true
     }
+    bldh-sles12sp4 = {
```
same

Bischoff commented on salt/controller/bashrc:
```diff
@@ -5,6 +5,7 @@ export SERVER="{{ grains.get('server') }}"
 {% if grains.get('proxy') | default(false, true) %}export PROXY="{{ grains.get('proxy') }}" {% else %}# no proxy defined {% endif %}
 {% if grains.get('client') | default(false, true) %}export CLIENT="{{ grains.get('client') }}"{% else %}# no client defined {% endif %}
 {% if grains.get('minion') | default(false, true) %}export MINION="{{ grains.get('minion') }}"{% else %}# no minion defined {% endif %}
+{% if grains.get('buildhost') | default(false, true) %}export BUILDHOST="{{ grains.get('buildhost') }}"{% else %}# no buildhost defined {% endif %}
```
let's continue copying `kvm_host` line below:
```
{% if grains.get('build_host') | default(false, true) %}export BUILD_HOST="{{ 
  grains.get('build_host') }}"{% else %}# no build host defined {% endif %}
```

with underscores + space

Bischoff commented on salt/top.sls:
```diff
@@ -18,6 +18,10 @@ base:
     - match: grain
     - minion
 
+  'roles:buildhost':
```
ok

calancha commented:
> could you please do a test to see if it works with a minion, but no build host? this machine should be optional.

Addressed the comments.  Now, testing if works as expected.

calancha commented:
> > could you please do a test to see if it works with a minion, but no build host? this machine should be optional.
> 
> Addressed the comments. Now, testing if works as expected.

OK, it works as expected.

rjmateus commented on modules/cucumber_testsuite/main.tf:
```diff
@@ -125,6 +125,26 @@ module "min-sles12sp4" {
+  base_configuration = module.base.configuration
+  product_version    = var.product_version
+  image              = lookup(local.images, "min-build", "sles12sp4")
+  name               = lookup(local.names, "min-build", "bldh-sles12")
```
name should be consistent with module name. Change to "min-build"

rjmateus commented on salt/top.sls:
```diff
@@ -18,6 +18,10 @@ base:
 
+  'roles:buildhost':
+    - match: grain
+    - buildhost
```
If I understand correctly if machine have role `buildhost` the states located in folder `buildhost` will be applied. I don't any place setting this role neither the needed salt states to be apply.

calancha replied:
Good catch.  That's not needed.  Deleted it.

calancha commented:
Got the OK from Moio at rockechat to merge, thus I am merging it.
